### PR TITLE
Implement SIMD narrow and widen

### PR DESCRIPTION
**Pull request by ngzhian**
These are the operations implemented:

- i8x16.narrow_i16x8_s(a: v128, b: v128) -> v128
- i8x16.narrow_i16x8_u(a: v128, b: v128) -> v128
- i16x8.narrow_i32x4_s(a: v128, b: v128) -> v128
- i16x8.narrow_i32x4_u(a: v128, b: v128) -> v128
- i16x8.widen_low_i8x16_s(a: v128) -> v128
- i16x8.widen_high_i8x16_s(a: v128) -> v128
- i16x8.widen_low_i8x16_u(a: v128) -> v128
- i16x8.widen_high_i8x16_u(a: v128) -> v128
- i32x4.widen_low_i16x8_s(a: v128) -> v128
- i32x4.widen_high_i16x8_s(a: v128) -> v128
- i32x4.widen_low_i16x8_u(a: v128) -> v128
- i32x4.widen_high_i16x8_u(a: v128) -> v128

This now passes test/core/simd/simd_conversions.wast.

### CODE REVIEW

File: interpreter/exec/simd.ml [hunk 1]
```diff
@@ -315,10 +331,46 @@ struct
...
     end)
 
+  let clamp x low high = min (max x low) high
+
+  (* Narrow two v128 into one v128 by using to_shape on both operands,
+   * concatenating them, clamping the element to between low and high,
+   * then of_shape to reconstruct a v128. *)
+  let narrow to_shape of_shape low high x y =
+    let xy = (to_shape x) @ (to_shape y) in
+    let narrow i = Int32.(clamp i (of_int low) (of_int high)) in
```
Comment (rossberg) [inline interpreter/exec/simd.ml:341]: Nit: directly pass `low` and `high` as int32 to avoid the conversions.

File: interpreter/exec/simd.ml [hunk 2]
```diff
@@ -315,10 +331,46 @@ struct
...
+    let narrow_u = narrow Rep.to_i16x8 Rep.of_i8x16 0 255
+  end
+
+  module I16x8_convert = struct
+    let narrow_s = narrow Rep.to_i32x4 Rep.of_i16x8 (-32768) 32767
+    let narrow_u = narrow Rep.to_i32x4 Rep.of_i16x8 0 65535
+
+    let unsigned_mask x = Int32.(logand x (of_int 0xff))
+    let widen take_or_drop f x =
+      Rep.of_i16x8 (List.map f ((take_or_drop 8) (Rep.to_i8x16 x)))
```
Comment (rossberg) [inline interpreter/exec/simd.ml:355]:
```suggestion
      Rep.of_i16x8 (List.map f (take_or_drop 8 (Rep.to_i8x16 x)))
```

File: interpreter/exec/simd.ml [hunk 3]
```diff
@@ -315,10 +331,46 @@ struct
...
+  end
+
   module I32x4_convert = struct
     let convert_using f v = Rep.of_i32x4 (List.map f (Rep.to_f32x4 v))
     let trunc_sat_f32x4_s = convert_using I32_convert.trunc_sat_f32_s
     let trunc_sat_f32x4_u = convert_using I32_convert.trunc_sat_f32_u
+
+    let unsigned_mask x = Int32.(logand x (of_int 0xffff))
+    let widen take_or_drop f x =
+      Rep.of_i32x4 (List.map f ((take_or_drop 4) (Rep.to_i16x8 x)))
```
Comment (rossberg) [inline interpreter/exec/simd.ml:369]:
```suggestion
      Rep.of_i32x4 (List.map f (take_or_drop 4 (Rep.to_i16x8 x)))
```

File: interpreter/exec/simd.ml [hunk 4]
```diff
@@ -315,10 +331,46 @@ struct
...
+
+  module I16x8_convert = struct
+    let narrow_s = narrow Rep.to_i32x4 Rep.of_i16x8 (-32768) 32767
+    let narrow_u = narrow Rep.to_i32x4 Rep.of_i16x8 0 65535
+
+    let unsigned_mask x = Int32.(logand x (of_int 0xff))
+    let widen take_or_drop f x =
+      Rep.of_i16x8 (List.map f ((take_or_drop 8) (Rep.to_i8x16 x)))
+    let widen_low_s = widen (Lib.List.take) (fun x -> x)
+    let widen_high_s = widen (Lib.List.drop) (fun x -> x)
```
Comment (rossberg) [inline interpreter/exec/simd.ml:357]:
```suggestion
    let widen_low_s = widen Lib.List.take Fun.id
    let widen_high_s = widen Lib.List.drop Fun.id
```

File: interpreter/exec/simd.ml [hunk 5]
```diff
@@ -315,10 +331,46 @@ struct
...
+  module I8x16_convert = struct
+    let narrow_s = narrow Rep.to_i16x8 Rep.of_i8x16 (-128) 127
+    let narrow_u = narrow Rep.to_i16x8 Rep.of_i8x16 0 255
+  end
+
+  module I16x8_convert = struct
+    let narrow_s = narrow Rep.to_i32x4 Rep.of_i16x8 (-32768) 32767
+    let narrow_u = narrow Rep.to_i32x4 Rep.of_i16x8 0 65535
+
+    let unsigned_mask x = Int32.(logand x (of_int 0xff))
```
Comment (rossberg) [inline interpreter/exec/simd.ml:353]:
```suggestion
    let unsigned_mask x = Int32.logand x 0xffl
```

File: interpreter/exec/simd.ml [hunk 6]
```diff
@@ -315,10 +331,46 @@ struct
...
+    let widen_low_u = widen (Lib.List.take) unsigned_mask
+    let widen_high_u = widen (Lib.List.drop) unsigned_mask
+  end
+
   module I32x4_convert = struct
     let convert_using f v = Rep.of_i32x4 (List.map f (Rep.to_f32x4 v))
     let trunc_sat_f32x4_s = convert_using I32_convert.trunc_sat_f32_s
     let trunc_sat_f32x4_u = convert_using I32_convert.trunc_sat_f32_u
+
+    let unsigned_mask x = Int32.(logand x (of_int 0xffff))
```
Comment (rossberg) [inline interpreter/exec/simd.ml:367]:
```suggestion
    let unsigned_mask x = Int32.logand x 0xffffl
```

File: interpreter/exec/simd.ml [hunk 7]
```diff
@@ -315,10 +331,46 @@ struct
...
+    let narrow_s = narrow Rep.to_i32x4 Rep.of_i16x8 (-32768) 32767
+    let narrow_u = narrow Rep.to_i32x4 Rep.of_i16x8 0 65535
+
+    let unsigned_mask x = Int32.(logand x (of_int 0xff))
+    let widen take_or_drop f x =
+      Rep.of_i16x8 (List.map f ((take_or_drop 8) (Rep.to_i8x16 x)))
+    let widen_low_s = widen (Lib.List.take) (fun x -> x)
+    let widen_high_s = widen (Lib.List.drop) (fun x -> x)
+    let widen_low_u = widen (Lib.List.take) unsigned_mask
+    let widen_high_u = widen (Lib.List.drop) unsigned_mask
```
Comment (rossberg) [inline interpreter/exec/simd.ml:359]: You could inline this as
```suggestion
    let widen_low_u = widen Lib.List.take (Int32.logand 0xffl)
    let widen_high_u = widen Lib.List.drop (Int32.logand 0xffl)
```

Or even simpler than using a function parameter: just have `widen` take a mask and pass in either 0xffffffffl or 0xffl.

File: interpreter/exec/simd.ml [hunk 8]
```diff
@@ -315,10 +331,46 @@ struct
...
     let trunc_sat_f32x4_s = convert_using I32_convert.trunc_sat_f32_s
     let trunc_sat_f32x4_u = convert_using I32_convert.trunc_sat_f32_u
+
+    let unsigned_mask x = Int32.(logand x (of_int 0xffff))
+    let widen take_or_drop f x =
+      Rep.of_i32x4 (List.map f ((take_or_drop 4) (Rep.to_i16x8 x)))
+    let widen_low_s = widen (Lib.List.take) (fun x -> x)
+    let widen_high_s = widen (Lib.List.drop) (fun x -> x)
+    let widen_low_u = widen (Lib.List.take) unsigned_mask
+    let widen_high_u = widen (Lib.List.drop) unsigned_mask
```
Comment (rossberg) [inline interpreter/exec/simd.ml:373]:
```suggestion
    let widen_low_s = widen Lib.List.take Fun.id
    let widen_high_s = widen Lib.List.drop Fun.id
    let widen_low_u = widen Lib.List.take (Int32.logand 0xffffl)
    let widen_high_u = widen Lib.List.drop (Int32.logand 0xffffl)
```

File: interpreter/exec/simd.ml (hunk 7 above)
Comment (ngzhian) [reply interpreter/exec/simd.ml]: Sg, made that change.

File: interpreter/exec/simd.ml (hunk 1 above)
Comment (ngzhian) [reply interpreter/exec/simd.ml]: Done (I never knew about the `l` suffix, thanks!)

File: interpreter/exec/simd.ml [hunk 9]
```diff
@@ -322,10 +338,43 @@ struct
...
+    let widen take_or_drop mask x =
+      Rep.of_i16x8 (List.map (Int32.logand mask) (take_or_drop 8 (Rep.to_i8x16 x)))
+    let widen_low_s = widen Lib.List.take 0xffffffffl
+    let widen_high_s = widen Lib.List.drop 0xffffffffl
+    let widen_low_u = widen Lib.List.take 0xffl
+    let widen_high_u = widen Lib.List.drop 0xffl
+  end
+
   module I32x4_convert = struct
     let convert_using f v = Rep.of_i32x4 (List.map f (Rep.to_f32x4 v))
```
Comment (rossberg) [inline interpreter/exec/simd.ml:368]: Nit: I'd remove the `_using` suffix on this one, since you don't have it anywhere else.

**Pull request merged**